modules/library/main/src/main/java/org/geotools/data/crs/ReprojectFeatureReader.java

```diff
@@ -109,18 +118,37 @@ public ReprojectFeatureReader(
             throw new NullPointerException("CoordinateSystem required");
         }
 
-        SimpleFeatureType type = reader.getFeatureType();
-        CoordinateReferenceSystem original =
-                type.getGeometryDescriptor().getCoordinateReferenceSystem();
+        this.originalType = reader.getFeatureType();
 
-        if (cs.equals(original)) {
+        if (!FeatureTypes.shouldReproject(originalType, cs)) {
             throw new IllegalArgumentException(
                     "CoordinateSystem " + cs + " already used (check before using wrapper)");
         }
 
-        this.schema = FeatureTypes.transform(type, cs);
+        this.schema = FeatureTypes.transform(originalType, cs);
+
         this.reader = reader;
-        transformer.setMathTransform(CRS.findMathTransform(original, cs, true));
+        this.transformers = new HashMap<>();
+        for (int i = 0; i < originalType.getDescriptors().size(); i++) {
+            if (originalType.getDescriptor(i) instanceof GeometryDescriptor) {
+                GeometryDescriptor descr = (GeometryDescriptor) originalType.getDescriptor(i);
+                CoordinateReferenceSystem original = descr.getCoordinateReferenceSystem();
+                if (CRS.isCompatible(cs, original)) {
```

Comment:
And if not compatible... it does nothing? Shouldn't it be throwing an exception instead? Please discuss :-)

Reply:
I was unsure myself but based this behaviour on the OGC standard. In "OGC® Web Feature Service 2.0 Interface Standard – With Corrigendum" 7.9.2.4.4 srsName parameter (https://docs.opengeospatial.org/is/09-025r2/09-025r2.html#95) 

"The optional srsName attribute may be used to assert a specific WFS-supported CRS transformation to be applied to the compatible (see 7.9.2.4.4.2) geometries of the features returned in a response document."

And then further in 7.9.2.4.4.2 Compatibility of coordinate reference systems:

"Coordinate reference systems that differ in dimensionality or that have no defined datum transformation between them may be incompatible for the purposes of geometry coordinate conversion. Where the request srsName references a CRS that is incompatible with the storage CRS of a geometry property, the service shall encode that geometry property using some other CRS identified with an srsName attribute on that property"

This seems to suggest to do nothing indeed for those geometries that are incompatible with the reprojection.

It seems that many "Inspire" schemas are requiring complex feature types with both a (primary) 2D and a (secondary) 3D geometry inside them. If an exception was thrown, any 2D reprojection applied to one of these schemas would fail. This is particularly problematic because in geoserver the implementations of some versions of WFS pass on a reprojection CRS in the Query always (even if no srsName was specified in the http request, it has something to do with the different ways different wfs versions handle crs encoding and axis order), so that all requests would fail.